core/shared/src/main/scala/scalaz/zio/IO.scala

```diff
@@ -786,6 +786,23 @@ object IO {
     in.foldLeft(point[E, mutable.Builder[B, M[B]]](cbf(in)))((io, b) => io.zipWith(fn(b))(_ += _))
       .map(_.result())
 
+  /**
+   * Evaluate the elements of a traversable data structure in parallel
+   * and collect the results.
+   *
+   * _Note_: ordering in the input collection is not preserved
+   */
+  def parTraverse[E, A, B, M[X] <: TraversableOnce[X]](
+    in: M[A]
+  )(fn: A => IO[E, B])(implicit cbf: CanBuildFrom[M[A], B, M[B]]): IO[E, M[B]] =
+    in.foldLeft[IO[E, Fiber[E, mutable.Builder[B, M[B]]]]](
+      point(cbf(in)).fork
+    )((iomb, a) =>
+      iomb.zipWith(fn(a).fork){ case (fmb, fb) => fmb.zipWith(fb){ case (mb, b) => mb += b }}
```

Comment:
You should not use impure code in `zipWith`. I think maybe it will work but it's very hard to reason about. Maybe for this initial version you could just use a list and then convert it using the builder.

Reply:
Also, you can in fact use `par` for this and it will have better behavior (in particular, if one fails, the result will fail and others will be interrupted).

Reply:
Thanks John. I've reverted to `par` as you suggested. Sorry I wasn't clear on what imperative style meant exactly, is this what you had in mind:
```scala
@tailrec def parTraverse_rec(as: Iterator[A], ioref: IO[E, IORef[List[B]]): IO[E, IORef[List[B]] =
  if (!as.hasNext) 
    point(ref)
  else {
    val (a, as') = (as.next, as.drop(1))
    parTraverse_rec(as', ioref.par(fn(a)).flatMap((ref, b) => ref.modify(b :: _) *> point(ref))
  }
}
parTraverse_rec(in.iterator, IORef[E, List[B]](Nil))... // unfold the list into an `M[B]`
```